This phylogeny shows evolutionary relationships of SARS-CoV-2 viruses from the ongoing novel coronavirus COVID-19 pandemic with a particular focus on Switzerland.
We aim to include as many genomes from Switzerland as possible.
In addition, we include virus genomes from other European countries and to lesser extent from the rest of the world to provide phylogenetic context.
This context is important to understand how different outbreaks are connected and how the virus might have spread around the globe.
We note, however, that direct linkage on the tree does not necessarily mean there exists a direct epidemiological link.
Sampling of viral genomes is very incomplete and heterogeneous and mutations occur on average only every two weeks.
Extended periods without mutations are therefore not uncommon and direct links or directionality cannot be inferred from the tree alone.

The data from Switzerland was contributed by the following groups:

 * [Computational Evolution group](https://bsse.ethz.ch/cevo) at D-BSSE of the ETHZ in Basel and the SIB Swiss Institute of Bioinformatics. Samples were provided by [Viollier AG](www.viollier.ch), sequenced on Illumina technology by the Genomic Facility Basel (lead [C. Beisel](https://bsse.ethz.ch/genomicsbasel)), and processed with the SIB Resource [V-Pipe viral assembly pipeline](https://cbg-ethz.github.io/V-pipe/sars-cov-2/) ([Ivan Topolsky, Pedro Ferreira, Philipp Jablonski, Susana Posada-​Céspedes, Niko Beerenwinkel](https://bsse.ethz.ch/cbg)).
 * University Hospital Basel, [Clinical Virology](https://www.unispital-basel.ch/ueber-uns/bereiche/medizinische-querschnittsfunktionen/kliniken-institute-abteilungen/labormedizin/ueber-uns/klinische-virologie/kontakt-und-oeffnungszeiten/) and [Clinical Microbiology](https://www.unispital-basel.ch/ueber-uns/bereiche/medizinische-querschnittsfunktionen/kliniken-institute-abteilungen/labormedizin/ueber-uns/klinische-bakteriologie-mykologie/kontakt-und-oeffnungszeiten/).
 * [Laboratory of Virology (University Hospitals of Geneva (HUG))](https://www.hug-ge.ch/laboratoire-virologie). Samples were collected from different cantons in Switzerland, sequenced on Illumina NovaSeq  (H2030 Genome Center) or HiSeq4000 (iGE3 Genomics Platform from the University of Geneva) instruments and processes by the Laboratory of Virology from HUG (Florian Laubscher, Samuel Cordey, Ana Rita Goncalves and Laurent Kaiser) and/or the H2030 Genome Center (Lorenzo Cerutti, Henri Pegeot, Melyssa Elies, Deborah Penet, Keith Harshman, Ioannis Xenarios, and Emmanouil Dermitzakis).
 * [Institute of Medical Virology at University of Zurich](https://www.virology.uzh.ch/de.html). Samples were provided by University Hospital Zurich, Triemli Hospital Zurich and Klinik Hirslanden Zurich, sequenced on Illumina technology and processed with VirMet viral metagenomic pipeline (Stefan Schmutz, Maryam Zaheri, Verena Kufner, Gabriela Ziltener, Jürg Böni, Michael Huber, Alexandra Trkola)

The analysis is maintained by
 * [Cornelius Roemer](https://neherlab.org) at Biozentrum of the University of Basel and the Swiss Institute of Bioinformatics
 * [Richard Neher](https://neherlab.org) at Biozentrum of the University of Basel and the Swiss Institute of Bioinformatics
 * [Sarah Nadeau](https://bsse.ethz.ch/cevo) at ETH Zürich and the Swiss Institute of Bioinformatics
 * [Tanja Stadler](https://bsse.ethz.ch/cevo) at ETH Zürich and the Swiss Institute of Bioinformatics


All data we use (Swiss and international) are deposited in GISAID.
We gratefully acknowledge the authors, originating and submitting laboratories of the genetic sequence and metadata made available through [GISAID](https://gisaid.org) on which this research is based. A full listing of all originating and submitting laboratories is available below. An attribution table is available by clicking on "Download Data" at the bottom of the page and then clicking on "Strain Metadata" in the resulting dialog box.
